text/0001-catalogs.md

```diff
@@ -0,0 +1,341 @@
+
+For example, the `foo` and `bar` packages of a monorepo may declare the same version of `react` in their `package.json` files.
+
+```json
+{
+  "name": "@monorepo/foo",
+  "dependencies": {
+    "react": "^18.2.0",
+  }
+}
+```
+
+```json
+{
+  "name": "@monorepo/bar",
+  "dependencies": {
+    "react": "^18.2.0",
+  }
+}
+```
+
+For smaller monorepos with a few packages, it's easy to manually ensure these versions stay in sync. For monorepos with hundreds of packages and many contributors, it becomes untenable to rely on code review to ensure every dependency has a single version for all packages.
+
+As a result, multiple versions of the same dependency appear over time. This can cause different flavors of surprising problems.
+
+- In projects that bundle dependencies, multiple versions inflate the size of the final result deployed to users.
+- Differing versions result in multiple copies that may not interact well at runtime, especially if features like [`Symbol()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Symbol) are used. For example, [React hooks will error if a component is rendered with a different copy of React in the same app](https://reactjs.org/warnings/invalid-hook-call-warning.html#duplicate-react).
+- For TypeScript, multiple copies of the same `@types` package causes compile errors from mismatching definitions. The compiler diagnostic for this is usually: *"argument of type `Foo` is not assignable to parameter of type `Foo`"*. For developers that have seen this before, they may realize this diagnostic is due to a dependency version mismatch. For developers new to TypeScript, *"`Foo` is not assignable to `Foo`"* is very confusing.
+
+While there are situations differing versions are intentional, this is more often accidental. Multiple differing versions arise from not reviewing `pnpm-lock.yaml` file changes or not searching for existing dependency specifiers before adding a new one. The later is typically unwritten convention in most monorepos.
+
+### Merge Conflict Resistance
+
+In addition to reducing the likelihood of multiple versions of the same dependency in a monorepo, the new `catalog:` protocol reduces merge conflicts. Any `package.json` files using the catalog protocol to declare a dependency do not need to be edited when changing the version of that dependency. This side steps `package.json` merge conflicts by avoiding edits to them in the first place.
+
+**Merge conflict resistance is the a primary motivator for introducing catalogs as a first-class feature to pnpm.** This is only possible through a new specifier protocol. See [_What kinds of merge conflicts are avoided?_](#what-kinds-of-merge-conflicts-are-avoided) for details.
+
+## Detailed Explanation
+
+### Configuring Catalogs
+
+Catalogs are configured on the root `package.json` of a workspace and available to all workspace packages.
```

Comment:
Configuring catalogs from the root `package.json` may make more sense than configuring templates from here. I decided to reincorporate this as as result.

There were some previous ideas on how to do this for templates here: https://github.com/pnpm/rfcs/pull/3#discussion_r1241479712 that we punted. But it may make sense to revisit this for catalogs alone.

Reply:
Going to hit "Resolve conversation" since this part seems to be non-contentious at this point.